core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2210,7 +2210,8 @@ private Comparison(RexNode ref, SqlKind kind, RexLiteral literal) {
     /** Creates a comparison, between a {@link RexInputRef} or {@link RexFieldAccess}
      * and a literal. */
     static Comparison of(RexNode e) {
-      return of(e, node -> RexUtil.isReferenceOrAccess(node, true));
+      return of(e, node -> RexUtil.isReferenceOrAccess(node, true)
+          || RexUtil.isDeterministic(node));
```

Comment:
Do you think we can use `RexUtilis#isConstant(RexNode node)` ? Or why a deterministic function should be simplified ? If that is the case, please also update the document.

Reply:
Is constant didn't satisfy my test, so I've updated javadoc.

Reply:
This is the core change of this PR, while the others are code refactoring, i would suggest to split the PR into 2 commits, one is for this change, and the other for the code refactoring.

Reply:
Thank you, but splitting in such a case is an extra effort for nothing because refactoring changes are very small and visible.